### PR TITLE
Add Adadelta optimizer

### PR DESCRIPTION
**Pull request by Lewuathe**
Adadelta is a optimization algorithm to overcome the sensitivity of learning rate selection of Adagrad.

see: http://www.matthewzeiler.com/wp-content/uploads/2017/07/googleTR2012.pdf

adadelta_cost
adadelta_accuracy

---
This change is [Reviewable](https://reviewable.io/reviews/pair-code/deeplearnjs/150)

### CODE REVIEW

**Comment by nsthorat** [comment]
Nice change! Will merge after a few tiny fixes, but the implementation looks correct :)

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/graph/session_test.ts, line 455 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-Kv2tb5DB8vbzoMAZqpU:-Kv2tb5DB8vbzoMAZqpV:b-yxmrzi) ([raw file](https://github.com/pair-code/deeplearnjs/blob/e21db2f3874577ec63468c3c3abaca2e538e0cca/src/graph/session_test.ts#L455)):*
> ```XML
>   });
> 
>   it('Safe mode math, math scope train does not throw', () => {
> ```

I totally missed this in the last reviews but the string names of these are not correct, can you rename this to "adelta"

---

*[src/graph/optimizers/adadelta_optimizer.ts, line 51 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-Kv2uOB4WXMy4IL8dWvx:-Kv2uOB4WXMy4IL8dWvy:bqx6ynr) ([raw file](https://github.com/pair-code/deeplearnjs/blob/e21db2f3874577ec63468c3c3abaca2e538e0cca/src/graph/optimizers/adadelta_optimizer.ts#L51)):*
> ```XML
>   }
> ```

remove blank line

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-:-Kv2udOyOmbubJ6P1j7D:bhe85l2)*

**Comment by Lewuathe** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/graph/session_test.ts, line 455 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-Kv2tb5DB8vbzoMAZqpU:-Kv4pGYAidF05YAqZ_7N:bcittr1) ([raw file](https://github.com/pair-code/deeplearnjs/blob/e21db2f3874577ec63468c3c3abaca2e538e0cca/src/graph/session_test.ts#L455)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

I totally missed this in the last reviews but the string names of these are not correct, can you rename this to "adelta"
</blockquote></details>

Sure will update.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150)*

**Comment by Lewuathe** [comment]
Review status: 1 of 3 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[src/graph/optimizers/adadelta_optimizer.ts, line 51 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-Kv2uOB4WXMy4IL8dWvx:-Kv7VHNFn3QOh9df2f9l:ba87zsj) ([raw file](https://github.com/pair-code/deeplearnjs/blob/e21db2f3874577ec63468c3c3abaca2e538e0cca/src/graph/optimizers/adadelta_optimizer.ts#L51)):*
<details><summary><i>Previously, nsthorat (Nikhil Thorat) wrote…</i></summary><blockquote>

remove blank line
</blockquote></details>

Do you think we can set `no-consecutive-blank-lines` of tslint setting?
I found several issues when I applied this setting to current codebase.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150)*

**Comment by nsthorat** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[src/graph/optimizers/adadelta_optimizer.ts, line 51 at r1](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-Kv2uOB4WXMy4IL8dWvx:-KvOP-AIYpUJC3EiVeee:b-o885pa) ([raw file](https://github.com/pair-code/deeplearnjs/blob/e21db2f3874577ec63468c3c3abaca2e538e0cca/src/graph/optimizers/adadelta_optimizer.ts#L51)):*
<details><summary><i>Previously, Lewuathe (Kai Sasaki) wrote…</i></summary><blockquote>

Do you think we can set `no-consecutive-blank-lines` of tslint setting?
I found several issues when I applied this setting to current codebase.
</blockquote></details>

Yup, we should do that. I'll make the change at some point soon, but you're welcome to make that change and send me a PR with the fixes if you'd like, no worries if you don't want to though :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150)*

**Comment by nsthorat** [comment]
:lgtm_strong:

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/pair-code/deeplearnjs/150#-:-KvOPBb8G_Iyfa4mP2M6:b-jbsf8z)*
